### PR TITLE
Unsolicited suggestion to make the Java CLI more extensible

### PR DESCRIPTION
**Pull request by berkeleysquare**
Here is a suggestion to the Java CLI. Instead of the giant switch, I instantiate the command by reflection, and I ask the command for the view instead of the big maps.

It's a HUGE pull request, but if you look at Ds3Cli.java and Main.java, you see where I am headed. 

I also collapsed several views that were similar to reuse one.

I renamed one or two commands to match the CommandValue. I kept the enum, but you could drop it and go to completely dynamic adding of commands -- that's a trade against keeping them honest.

No doubt there's significant prettying to be done. But do y'all like it and think it worth pursuing? I thought it facilitated new commands (a bridge toward autogen?) and would be easier to internationalize.

All the Unit tests passed, so I think it is transparent to users.

### CODE REVIEW

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Ds3Cli.java [hunk 1]
```diff
@@ -168,111 +88,36 @@ public CommandResponse getCommandHelp() throws Exception {
...
-            }
-            case RECLAIM_CACHE: {
-                return new ReclaimCache(this.ds3Provider,  this.fileUtils);
-            }
-            default: {
-                LOG.error("Unimplemented command: " + command.name());
-                throw new CommandException(("Unimplemented command: " + command.name()));
+        final String commandName = this.args.getCommand();
+
+        String commandCamel = CaseFormat.UPPER_UNDERSCORE.to(CaseFormat.UPPER_CAMEL, commandName.toString());
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Ds3Cli.java]: Missing `final`

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Ds3Cli.java [hunk 2]
```diff
@@ -168,111 +88,36 @@ public CommandResponse getCommandHelp() throws Exception {
...
+
+        Iterator<CliCommand> implementations = getAllCommands();
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Ds3Cli.java]: Missing `final`

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Ds3Cli.java [hunk 3]
```diff
@@ -168,111 +88,36 @@ public CommandResponse getCommandHelp() throws Exception {
...
+        while (implementations.hasNext()) {
+            CliCommand implementation = implementations.next();
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Ds3Cli.java]: Missing `final`

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Ds3Cli.java [hunk 4]
```diff
@@ -168,111 +88,36 @@ public CommandResponse getCommandHelp() throws Exception {
...
+            String className = implementation.getClass().getSimpleName();
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Ds3Cli.java]: Missing `final`

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Ds3Cli.java [hunk 5]
```diff
@@ -168,111 +88,36 @@ public CommandResponse getCommandHelp() throws Exception {
...
+            if (className.equalsIgnoreCase(commandCamel)) {
+                return implementation.withProvider(this.ds3Provider, this.fileUtils);
             }
         }
+         throw new CommandException("No command class: " + commandName);
+    }
+
+    private String listAllCommands() {
+        String commands = "Installed Commands: ";
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Ds3Cli.java]: Use a StringBuilder here instead of a String

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Ds3Cli.java [hunk 6]
```diff
@@ -168,111 +88,36 @@ public CommandResponse getCommandHelp() throws Exception {
...
+        Iterator<CliCommand> implementations = getAllCommands();
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Ds3Cli.java]: Missing `final`

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Ds3Cli.java [hunk 7]
```diff
@@ -168,111 +88,36 @@ public CommandResponse getCommandHelp() throws Exception {
...
+        while (implementations.hasNext()) {
+            CliCommand implementation = implementations.next();
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Ds3Cli.java]: Missing `final`

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Ds3Cli.java [hunk 8]
```diff
@@ -168,111 +88,36 @@ public CommandResponse getCommandHelp() throws Exception {
...
+            String unCamel = CaseFormat.UPPER_CAMEL.to(CaseFormat.LOWER_UNDERSCORE, implementation.getClass().getSimpleName());
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Ds3Cli.java]: Missing `final`

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/CliCommand.java [hunk 9]
```diff
@@ -29,8 +31,23 @@
 
 public abstract class CliCommand<T extends Result> implements Callable<T> {
 
-    private final Ds3Provider ds3Provider;
-    private final FileUtils fileUtils;
+    private Ds3Provider ds3Provider;
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/CliCommand.java]: Why did you remove the final decorator from these two variables?

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/DeleteBucket.java [hunk 10]
```diff
@@ -34,15 +34,15 @@
...
 
     private static final Logger LOG = LoggerFactory.getLogger(DeleteBucket.class);
 
     private String bucketName;
     private boolean force;
 
-    public DeleteBucket(final Ds3Provider provider, final FileUtils fileUtils) {
-        super(provider, fileUtils);
+    public DeleteBucket()
+    {
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/DeleteBucket.java]: The `{` should be at the end of the previous line.

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/DeleteFolder.java [hunk 11]
```diff
@@ -17,20 +17,23 @@
...
 
-public class DeleteFolder extends CliCommand<DeleteResult> {
+public class DeleteFolder extends CliCommand<DefaultResult> {
 
     private String bucketName;
     private String folderName;
 
-    public DeleteFolder(final Ds3Provider provider, final FileUtils fileUtils) { super(provider, fileUtils); }
+    public DeleteFolder()
+    {
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/DeleteFolder.java]: The `{` should be at the end of the previous line

**Comment by rpmoore** [comment]
Each of the views that are attached to the commands should be made static

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Ds3Cli.java [hunk 12]
```diff
@@ -168,111 +88,36 @@ public CommandResponse getCommandHelp() throws Exception {
...
+            commands += unCamel  + ", ";
+        }
+        return commands;
+    }
+
+    private Iterator<CliCommand> getAllCommands()
+    {
+        ServiceLoader<CliCommand> loader =
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Ds3Cli.java]: Missing `final`

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/CliCommand.java [hunk 13]
```diff
@@ -29,8 +31,23 @@
...
+    private FileUtils fileUtils;
+
+    protected final View<T> cliView = (View<T>) new DefaultView();
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/CliCommand.java]: Do you have to cast this with the `(View<T>)`?  So long as DefaultView is implemented to return a T that extends Result, you shouldn't have to cast it...

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/CliCommand.java [hunk 14]
```diff
@@ -51,8 +68,17 @@ protected FileUtils getFileUtils() {
...
     public abstract CliCommand init(final Arguments args) throws Exception;
 
+
     // help for '--help' command
-    public String getLongHelp(CommandValue command) {
+    public String getLongHelp(final String command) {
         return CommandHelpText.getHelpText(command) ;
     }
+
+    public View getView(final ViewType viewType) {
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/CliCommand.java]: Shouldn't getView return View<T>?

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/CliCommand.java (hunk 9 above)
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/CliCommand.java]: Oh I see why you removed it.

File: ds3-cli-integration/src/main/resources/META-INF/services/com.spectralogic.ds3cli.command.CliCommand [hunk 15]
```diff
@@ -0,0 +1,41 @@
+# CLI Command implementatiosn in this jar
```
Comment (rpmoore) [inline ds3-cli-integration/src/main/resources/META-INF/services/com.spectralogic.ds3cli.command.CliCommand]: Do you need this in the integration jar file?  Shouldn't the ServiceProvider look this up in the other jar file?

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/views/cli/TableView.java [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+import com.spectralogic.ds3cli.models.Result;
+
+public abstract class TableView<T extends Result> implements View<T> {
+
+    protected ASCIITableHeader[] header;
+    protected int columnCount;
+
+    public abstract String render(final T result);
+
+    public void initTable(ImmutableList<String> columnHeads) {
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/views/cli/TableView.java]: Missing `final`

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/views/cli/TableView.java [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+        this.columnCount = columnHeads.size();
+
+        // create the header
+        this.header = new ASCIITableHeader[this.columnCount];
+        for (int i = 0; i < this.columnCount; i++) {
+            header[i] = (new ASCIITableHeader(columnHeads.get(i), ASCIITable.ALIGN_LEFT));
+        }
+    }
+
+    public void setTableDataAlignment(ImmutableList<Integer> columnAlign) {
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/views/cli/TableView.java]: Missing `final`

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Ds3Cli.java [hunk 18]
```diff
@@ -155,6 +69,12 @@ public CommandResponse call() throws Exception {
     }
 
     public CommandResponse getCommandHelp() throws Exception {
+        if (this.args.getCommand().equalsIgnoreCase("LIST_COMMANDS"))
+        {
```
Comment (berkeleysquare) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Ds3Cli.java]: { to line above

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Ds3Cli.java [hunk 19]
```diff
@@ -168,111 +87,38 @@ public CommandResponse getCommandHelp() throws Exception {
...
+            commands.append(CaseFormat.UPPER_CAMEL.to(CaseFormat.LOWER_UNDERSCORE, implementation.getClass().getSimpleName()));
+            if (implementations.hasNext()) {
+                commands.append(", ");
             }
         }
+        return commands.toString();
+    }
+
+    private Iterator<CliCommand> getAllCommands()
+    {
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Ds3Cli.java]: The `{` character should be at the end of the previous line

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/views/cli/TableView.java [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+
+    public abstract String render(final T result);
+
+    public void initTable(final ImmutableList<String> columnHeads) {
+        this.columnCount = columnHeads.size();
+
+        // create the header
+        this.header = new ASCIITableHeader[this.columnCount];
+        for (int i = 0; i < this.columnCount; i++) {
+            header[i] = (new ASCIITableHeader(columnHeads.get(i), ASCIITable.ALIGN_LEFT));
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/views/cli/TableView.java]: The outer parens are not required.

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/views/json/DefaultView.java~ [hunk 21]
```diff
@@ -1,5 +1,5 @@
 /*
- * *****************************************************************************
+ * ******************************************************************************
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/views/json/DefaultView.java~]: It looks like this file was unintentionally checked in.
